Typescript classes for component options

danice opened the pull request:
## Proposed changes
Add classes for component options. This will give typescript info of the available properties.
As an idea the options class can have a constructor of this kind:
```
constructor(init?: Partial<AutocompleteOptions>)
  {
    Object.assign(this, {..._defaults, ...init});         
  }
```
this will allow initialisations as 

`new AutocompleteOptions({minLength: 2})`

## Screenshots (if appropriate) or codepen:
<!-- Add supplemental screenshots or code examples. Look for a codepen template in our **[CONTRIBUTING document](https://github.com/materializecss/materialize/blob/master/CONTRIBUTING.md)**. -->

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to change).

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the **[CONTRIBUTING document](https://github.com/materializecss/materialize/blob/main/CONTRIBUTING.md)**.
- [x] My commit messages follows the [conventional commit format](https://github.com/materializecss/materialize/blob/main/CONTRIBUTING.md#submitting-your-pull-request)
- [ ] My change requires a change to the documentation, and updated it accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

wuda-io commented:
Yes this is awesome and a nice bonus! I am not sure if this way is the way to go.
We should consider using an interface for the Options
and a seperate class for the default values...
like this:
```typescript
interface ACOptions {
   param1: string,
   optionalParam2?: number
}

const defaults: ACOptions = {
   param1: 'default-string'
}

export class AC extends Component {
  constructor (el, options: ACOptions) {
     this._options = {...defaults, ...options};
  }
}
```
Benefit here would be we can set required params and optional params.
I dont know about how the <Partial> flag in your code works.

danice commented on src/autocomplete.ts:
```diff
@@ -1,20 +1,31 @@
   isMultiSelect: false,
-  onSearch: function(text, autocomplete) {
-    const filteredData = autocomplete.options.data.filter(item => {
+  onSearch: (text, autocomplete, options) => {
```
as options is protected, it has to be passed as parameter for onSearch

danice commented on src/autocomplete.ts:
```diff
@@ -1,20 +1,31 @@
-let _defaults = {
+export interface AutocompleteOptions {
+  data?: any[];
+  onAutocomplete?(selectedValue: any): void;
```
this is called as
`this.options.onAutocomplete.call(this, this.selectedValues);`
so it does not need Autocomplete as parameter. 
But other callback as onSearch are called like
`this.options.onSearch(this.el.value, this, this.options);`
so this needs to have the component as a callback parameter. 

I would propose to unify this, and always having the component as parameter for callback methods..
What do you think?

danice commented:
I completely agree that using interfaces is much better for the options. Changed.
I have added some comments about the different ways callback methods are called...
